File: solutions/typescript/two-bucket/1/two-bucket.ts
export class TwoBucket {
  private _moves: number | null = null;
  private _goalBucket: string = '';
  private _otherBucketValue: number = 0;
  private _solutionFound: boolean = false;

  constructor(
    private bucketOneCap: number,
    private bucketTwoCap: number,
    private goal: number,
    private startBucket: string
  ) {
    this.solve();
  }

  private calculateGCD(a: number, b: number): number {
    return b === 0 ? a : this.calculateGCD(b, a % b);
  }

  private isForbiddenState(b1: number, b2: number): boolean {
    // Forbidden: starting bucket empty AND other bucket full
    if (this.startBucket === 'one') {
      return b1 === 0 && b2 === this.bucketTwoCap;
    } else {
      return b2 === 0 && b1 === this.bucketOneCap;
    }
  }

  private solve(): void {
    // Check if goal is reachable
    if (this.goal > Math.max(this.bucketOneCap, this.bucketTwoCap)) {
      this._solutionFound = false;
      return;
    }

    const gcd = this.calculateGCD(this.bucketOneCap, this.bucketTwoCap);
    if (this.goal % gcd !== 0) {
      this._solutionFound = false;
      return;
    }

    const queue: { b1: number; b2: number; moves: number }[] = [];
    const visited = new Set<string>();

    // Initial state: fill the starting bucket (this is move 1)
    let initialB1: number, initialB2: number;
    if (this.startBucket === 'one') {
      initialB1 = this.bucketOneCap;
      initialB2 = 0;
    } else {
      initialB1 = 0;
      initialB2 = this.bucketTwoCap;
    }

    queue.push({ b1: initialB1, b2: initialB2, moves: 1 });
    visited.add(`${initialB1},${initialB2}`);

    while (queue.length > 0) {
      const current = queue.shift()!;

      // Check if goal is reached
      if (current.b1 === this.goal) {
        this._moves = current.moves;
        this._goalBucket = 'one';
        this._otherBucketValue = current.b2;
        this._solutionFound = true;
        return;
      }
      if (current.b2 === this.goal) {
        this._moves = current.moves;
        this._goalBucket = 'two';
        this._otherBucketValue = current.b1;
        this._solutionFound = true;
        return;
      }

      // Generate next states
      const nextStates = this.getNextStates(current.b1, current.b2, current.moves);

      for (const next of nextStates) {
        const key = `${next.b1},${next.b2}`;
        if (!visited.has(key)) {
          visited.add(key);
          queue.push(next);
        }
      }
    }

    this._solutionFound = false;
  }

  private getNextStates(
    b1: number,
    b2: number,
    moves: number
  ): { b1: number; b2: number; moves: number }[] {
    const nextStates: { b1: number; b2: number; moves: number }[] = [];
    const nextMoves = moves + 1;

    // Action 1: Fill bucket one
    if (b1 < this.bucketOneCap) {
      const newState = { b1: this.bucketOneCap, b2, moves: nextMoves };
      if (!this.isForbiddenState(newState.b1, newState.b2)) {
        nextStates.push(newState);
      }
    }

    // Action 2: Fill bucket two
    if (b2 < this.bucketTwoCap) {
      const newState = { b1, b2: this.bucketTwoCap, moves: nextMoves };
      if (!this.isForbiddenState(newState.b1, newState.b2)) {
        nextStates.push(newState);
      }
    }

    // Action 3: Empty bucket one
    if (b1 > 0) {
      const newState = { b1: 0, b2, moves: nextMoves };
      if (!this.isForbiddenState(newState.b1, newState.b2)) {
        nextStates.push(newState);
      }
    }

    // Action 4: Empty bucket two
    if (b2 > 0) {
      const newState = { b1, b2: 0, moves: nextMoves };
      if (!this.isForbiddenState(newState.b1, newState.b2)) {
        nextStates.push(newState);
      }
    }

    // Action 5: Pour bucket one into bucket two
    if (b1 > 0 && b2 < this.bucketTwoCap) {
      const pour = Math.min(b1, this.bucketTwoCap - b2);
      const newState = { b1: b1 - pour, b2: b2 + pour, moves: nextMoves };
      if (!this.isForbiddenState(newState.b1, newState.b2)) {
        nextStates.push(newState);
      }
    }

    // Action 6: Pour bucket two into bucket one
    if (b2 > 0 && b1 < this.bucketOneCap) {
      const pour = Math.min(b2, this.bucketOneCap - b1);
      const newState = { b1: b1 + pour, b2: b2 - pour, moves: nextMoves };
      if (!this.isForbiddenState(newState.b1, newState.b2)) {
        nextStates.push(newState);
      }
    }

    return nextStates;
  }

  public moves(): number {
    if (!this._solutionFound || this._moves === null) {
      throw new Error('No solution found');
    }
    return this._moves;
  }

  public get goalBucket(): string {
    if (!this._solutionFound) {
      throw new Error('No solution found');
    }
    return this._goalBucket;
  }

  public get otherBucket(): number {
    if (!this._solutionFound) {
      throw new Error('No solution found');
    }
    return this._otherBucketValue;
  }
}

File: solutions/java/two-bucket/1/src/main/java/TwoBucket.java
import java.util.*;

class TwoBucket {
    private final Result result;

    TwoBucket(int bucketOneCap, int bucketTwoCap, int desiredLiters, String startBucket) {
        // If the desired amount is larger than both buckets, it's impossible
        if (desiredLiters > bucketOneCap && desiredLiters > bucketTwoCap) {
            throw new UnreachableGoalException();
        }

        Queue<int[]> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();

        int b1 = 0, b2 = 0;
        
        // First action is always filling the starting bucket
        if (startBucket.equals("one")) {
            b1 = bucketOneCap;
        } else {
            b2 = bucketTwoCap;
        }
        
        queue.add(new int[]{b1, b2, 1});
        visited.add(b1 + "," + b2);
        
        // Rule: "you may not end up in a state after any action where the 
        // starting bucket is empty and the other bucket is full."
        String forbiddenState = startBucket.equals("one") ? "0," + bucketTwoCap : bucketOneCap + ",0";

        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int c1 = curr[0];
            int c2 = curr[1];
            int moves = curr[2];

            // Check if we've reached the goal
            if (c1 == desiredLiters || c2 == desiredLiters) {
                String finalBucket = (c1 == desiredLiters) ? "one" : "two";
                int otherBucket = (c1 == desiredLiters) ? c2 : c1;
                this.result = new Result(moves, finalBucket, otherBucket);
                return;
            }

            // Generate all possible next states
            List<int[]> nextStates = new ArrayList<>();
            
            // 1. Fill bucket 1
            nextStates.add(new int[]{bucketOneCap, c2});
            // 2. Fill bucket 2
            nextStates.add(new int[]{c1, bucketTwoCap});
            // 3. Empty bucket 1
            nextStates.add(new int[]{0, c2});
            // 4. Empty bucket 2
            nextStates.add(new int[]{c1, 0});
            
            // 5. Pour bucket 1 into bucket 2
            int pour1to2 = Math.min(c1, bucketTwoCap - c2);
            nextStates.add(new int[]{c1 - pour1to2, c2 + pour1to2});
            
            // 6. Pour bucket 2 into bucket 1
            int pour2to1 = Math.min(c2, bucketOneCap - c1);
            nextStates.add(new int[]{c1 + pour2to1, c2 - pour2to1});

            // Process valid next states
            for (int[] next : nextStates) {
                String stateStr = next[0] + "," + next[1];
                // Avoid cycles and forbidden states
                if (!visited.contains(stateStr) && !stateStr.equals(forbiddenState)) {
                    visited.add(stateStr);
                    queue.add(new int[]{next[0], next[1], moves + 1});
                }
            }
        }

        // If the queue is exhausted and goal wasn't reached
        throw new UnreachableGoalException();
    }

    Result getResult() {
        return result;
    }
}